// @nilecare/web-dashboard/src/pages/Dashboards/AdminDashboard.tsx
/**
 * Admin Dashboard
 * System management, user management, reports
 */

import React, { useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import {
  Grid,
  Card,
  CardContent,
  Typography,
  Box,
  Button,
  Avatar,
  Chip,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
  CircularProgress,
} from '@mui/material';
import {
  People,
  LocalHospital,
  Assessment,
  Security,
  TrendingUp,
  Settings,
  Warning,
  BusinessCenter,
} from '@mui/icons-material';
import { useAuth } from '../../contexts/AuthContext';
import BusinessServiceCard from '../../components/BusinessServiceCard';
import { API_ENDPOINTS } from '../../config/api.config';

const StatCard: React.FC<{
  title: string;
  value: string | number;
  icon: React.ReactNode;
  color: string;
  change?: string;
}> = ({ title, value, icon, color, change }) => (
  <Card>
    <CardContent>
      <Box display="flex" alignItems="center" justifyContent="space-between">
        <Box>
          <Typography color="textSecondary" variant="body2">
            {title}
          </Typography>
          <Typography variant="h4" fontWeight="bold">
            {value}
          </Typography>
          {change && <Chip label={change} size="small" color="success" sx={{ mt: 1 }} />}
        </Box>
        <Avatar sx={{ bgcolor: color, width: 56, height: 56 }}>
          {icon}
        </Avatar>
      </Box>
    </CardContent>
  </Card>
);

const AdminDashboard: React.FC = () => {
  const { user } = useAuth();
  const navigate = useNavigate();

  const [loading, setLoading] = useState(true);
  const [stats, setStats] = useState({ totalUsers: 0, facilities: 0, systemHealth: 0, securityAlerts: 0 });
  const [systemStats, setSystemStats] = useState<any[]>([]);
  const [recentUsers, setRecentUsers] = useState<any[]>([]);

  useEffect(() => {
    const fetchData = async () => {
      try {
        setLoading(true);

        // Fetch dashboard stats
        const statsRes = await fetch(API_ENDPOINTS.dashboardStats);
        const statsData = await statsRes.json();
        if (statsData.success) {
          setStats({
            totalUsers: statsData.data.totalUsers || 0,
            facilities: 18, // Mock
            systemHealth: 98,
            securityAlerts: 3
          });

          // Set system stats from DB data
          setSystemStats([
            { metric: 'Active Users', value: statsData.data.totalUsers || 0, status: 'Normal' },
            { metric: 'Total Patients', value: statsData.data.totalPatients || 0, status: 'Normal' },
            { metric: 'Pending Labs', value: statsData.data.pendingLabs || 0, status: 'Normal' },
            { metric: 'Low Stock Items', value: statsData.data.lowStockItems || 0, status: statsData.data.lowStockItems > 30 ? 'Attention' : 'Good' },
          ]);
        }

        // Mock recent users for now
        setRecentUsers([
          { name: 'Dr. Ahmed Hassan', role: 'Doctor', status: 'Active', lastLogin: '2 hours ago' },
          { name: 'Nurse Sarah Ali', role: 'Nurse', status: 'Active', lastLogin: '30 min ago' },
          { name: 'Admin User', role: 'Admin', status: 'Active', lastLogin: 'Now' },
        ]);

      } catch (error) {
        console.error('Error fetching admin dashboard data:', error);
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, []);

  if (loading) {
    return (
      <Box display="flex" justifyContent="center" alignItems="center" minHeight="400px">
        <CircularProgress size={60} />
      </Box>
    );
  }

  return (
    <Box>
      <Box mb={3}>
        <Typography variant="h4" gutterBottom fontWeight="bold">
          Admin Dashboard - {user?.username || user?.first_name} 👨‍💼
        </Typography>
        <Typography variant="body1" color="text.secondary">
          System overview - {stats.totalUsers} active users
        </Typography>
      </Box>

      {/* Statistics - REAL DATA */}
      <Grid container spacing={3} mb={3}>
        <Grid item xs={12} sm={6} md={3}>
          <StatCard
            title="Total Users"
            value={stats.totalUsers}
            icon={<People />}
            color="#1976d2"
            change="Database"
          />
        </Grid>
        <Grid item xs={12} sm={6} md={3}>
          <StatCard
            title="Facilities"
            value={stats.facilities}
            icon={<LocalHospital />}
            color="#2e7d32"
          />
        </Grid>
        <Grid item xs={12} sm={6} md={3}>
          <StatCard
            title="System Health"
            value={`${stats.systemHealth}%`}
            icon={<TrendingUp />}
            color="#00a86b"
            change="Uptime"
          />
        </Grid>
        <Grid item xs={12} sm={6} md={3}>
          <StatCard
            title="Security Alerts"
            value={stats.securityAlerts}
            icon={<Security />}
            color="#d32f2f"
          />
        </Grid>
      </Grid>

      <Grid container spacing={3}>
        {/* System Stats */}
        <Grid item xs={12} md={6}>
          <Card>
            <CardContent>
              <Typography variant="h6" fontWeight="bold" mb={2}>
                ⚙️ System Statistics
              </Typography>
              <Table size="small">
                <TableHead>
                  <TableRow>
                    <TableCell><strong>Metric</strong></TableCell>
                    <TableCell><strong>Value</strong></TableCell>
                    <TableCell><strong>Status</strong></TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {systemStats.map((stat, idx) => (
                    <TableRow key={idx}>
                      <TableCell>{stat.metric}</TableCell>
                      <TableCell>{stat.value}</TableCell>
                      <TableCell>
                        <Chip
                          label={stat.status}
                          size="small"
                          color={stat.status === 'Good' || stat.status === 'Normal' ? 'success' : 'warning'}
                        />
                      </TableCell>
                    </TableRow>
                  ))}
                </TableBody>
              </Table>
            </CardContent>
          </Card>
        </Grid>

        {/* Recent Users */}
        <Grid item xs={12} md={6}>
          <Card>
            <CardContent>
              <Typography variant="h6" fontWeight="bold" mb={2}>
                👥 Recent User Activity
              </Typography>
              <Table size="small">
                <TableHead>
                  <TableRow>
                    <TableCell><strong>User</strong></TableCell>
                    <TableCell><strong>Role</strong></TableCell>
                    <TableCell><strong>Status</strong></TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {recentUsers.map((u, idx) => (
                    <TableRow key={idx}>
                      <TableCell>
                        <Box display="flex" alignItems="center">
                          <Avatar sx={{ mr: 1, width: 32, height: 32 }}>{u.name.charAt(0)}</Avatar>
                          <Box>
                            <Typography variant="body2">{u.name}</Typography>
                            <Typography variant="caption" color="text.secondary">
                              {u.lastLogin}
                            </Typography>
                          </Box>
                        </Box>
                      </TableCell>
                      <TableCell>{u.role}</TableCell>
                      <TableCell>
                        <Chip
                          label={u.status}
                          size="small"
                          color={u.status === 'Active' ? 'success' : 'default'}
                        />
                      </TableCell>
                    </TableRow>
                  ))}
                </TableBody>
              </Table>
              <Button fullWidth variant="outlined" sx={{ mt: 2 }}>
                View All Users
              </Button>
            </CardContent>
          </Card>
        </Grid>

        {/* Quick Actions */}
        <Grid item xs={12}>
          <Card>
            <CardContent>
              <Typography variant="h6" fontWeight="bold" mb={2}>
                🚀 Admin Actions
              </Typography>
              <Grid container spacing={2}>
                <Grid item xs={6} sm={3}>
                  <Button 
                    fullWidth 
                    variant="contained" 
                    startIcon={<People />}
                    onClick={() => navigate('/dashboard/admin/users')}
                  >
                    Manage Users
                  </Button>
                </Grid>
                <Grid item xs={6} sm={3}>
                  <Button 
                    fullWidth 
                    variant="contained" 
                    startIcon={<Assessment />} 
                    color="secondary"
                    onClick={() => alert('Reports module - Coming soon!')}
                  >
                    Reports
                  </Button>
                </Grid>
                <Grid item xs={6} sm={3}>
                  <Button 
                    fullWidth 
                    variant="contained" 
                    startIcon={<Settings />} 
                    color="info"
                    onClick={() => navigate('/dashboard/settings')}
                  >
                    Settings
                  </Button>
                </Grid>
                <Grid item xs={6} sm={3}>
                  <Button 
                    fullWidth 
                    variant="contained" 
                    startIcon={<Security />} 
                    color="error"
                    onClick={() => alert('Security Dashboard - Coming soon!')}
                  >
                    Security
                  </Button>
                </Grid>
              </Grid>
            </CardContent>
          </Card>
        </Grid>

        {/* Business Services Integration */}
        <Grid item xs={12}>
          <BusinessServiceCard onNavigate={(path) => navigate(path)} />
        </Grid>
      </Grid>
    </Box>
  );
};

export default AdminDashboard;
